File: generate.rb
#!/usr/bin/env ruby
require 'json'
require 'faraday'

def get_servicemap
  download = Faraday.get 'https://awsiamconsole.s3.amazonaws.com/iam/assets/js/bundles/policies.js'
  File.write './policies.js', download.body
end

def parse_servicemap
  policies = File.read('./policies.js')
  # Extract the ServiceMap
  index = policies.index('serviceMap:') + 'serviceMap:'.length
  chars = policies[index..-1].split ''
  # Pull the following object
  obj = []
  lb=rb=0
  chars.each do |c|
    lb += 1 if c== '{'
    rb += 1 if c== '}'
    obj.push c
    break if lb>0 and lb==rb
  end
  # Turn Javascript Object into JSON (FML -- can't aws just give us the JSON?)
  servicemap = obj.join.gsub(/\s/, '')
  # Cleanup Exceptions
  servicemap.gsub!(/HasResource:\!([0-9]+)/, 'HasResource:"!\1"')
  servicemap.gsub!(/ARNRegex:("[^"]*")/, 'ARNRegex:"REMOVED"')
  # turn objects into strings e.g. test:[] => "test":[]
  servicemap.gsub!(/([a-zA-Z0-9_]+):([\[{0-9"])/, '"\1":\2')
  # verify that our json is correct
  JSON.load(servicemap)
  File.write('servicemap.json', servicemap)
end


get_servicemap
parse_servicemap

File: generate.js
#!/usr/bin/env node
var exports = module.exports = {};


var https = require('https');
var fs    = require('fs') ;

exports.get_servicemap = function(callback) {
  var url = 'https://awspolicygen.s3.amazonaws.com/js/policies.js';
  https.get(url, (response) => {
    var body = '';
    response.on('data', (d) => { body += d; });
    response.on('end', () => {
      fs.writeFile(`${__dirname}/policies.js`, body, (error) => {
        if (error) console.error(error);
        else callback(body);
      });
    });
    response.on('error', (error) => {
      console.error('Error Downloading Policies.js', errormessage)
    });
  });
}

var load_policies = function (callback) {
  fs.readFile(`${__dirname}/policies.js`, 'utf8', (err, data) => {
    callback(data)
  });
}

exports.parse_servicemap = function(callback) {
  load_policies( (policies) => {
    var servicemap = policies.substring(policies.indexOf('{'));
    // verify that our json is correct
    servicemap = JSON.stringify(JSON.parse(servicemap), null, 2);
    fs.writeFile(`${__dirname}/servicemap.json`, servicemap, (err) => {
      if (err) console.error(error);
      callback(servicemap);
    });
  });
}

exports.run = function(callback) {
  exports.get_servicemap( (data) => {
    exports.parse_servicemap(callback)
  });
}



if (!module.parent) {
  // only run if this is the main module
  exports.run(console.log)
}
